### PR TITLE
Rename the generated bnfc to ast instead of Syntax based on the discussion #3193 (comment)

### LINKED ISSUE #3001: Refactoring and cleanup
## Overview
This issue is an epic to collect a list of proposals for code refactoring, file rearrangement or cleanups with the purpose to unify the design and strategy in organization the source code in the project.

### Restriction to type of changes
The important restriction is that changes should not impact functionality. Exceptions must be well justified.

### How to use this issue
Proposals will be collected in the description of the issue and updated when new items are added or status of existing is changed. 
From this issue multiple PR's will be spawn and linked to the issue.

### Proposals
- [ ] __Create _SourceEnclosing_ helper and remove LogSource and Metrics.Source__
  With simple macro we can eliminate creating LogSource on many places. We already have macro introduced here https://github.com/rchain/rchain/pull/1670#discussion_r227012153.
  This can be improved without the need to have implicit parameter for logger functions although with Metrics it's little more complex.
  Little complication is for shared project itself because macros must be compiled separately. One solution is to create subproject.
  https://www.scala-sbt.org/1.x/docs/Macro-Projects.html
  
- [ ] __Remove _metric_ cache, Kamon already keep internal map__
  https://github.com/rchain/rchain/blob/66b93e3/node/src/main/scala/coop/rchain/node/diagnostics/effects/package.scala#L117
  See Kamon source _MetricRegistry#lookupMetric_. Maybe _timer_ should be in bracket to ensure stop (?).

- [ ] __Replace custom `Time` with cats `Timer`__
  We can define custom [Clock](https://typelevel.org/cats-effect/datatypes/clock.html) and use cats `Timer` object. Time is introduced here https://github.com/rchain/rchain/pull/1558.
  Examples of `Timer` in tests and virtual (logical) time.
  https://blog.softwaremill.com/time-traveling-in-tests-with-cats-effect-b22084f6a89
  https://monix.io/api/3.0/monix/execution/schedulers/TestScheduler.html
    
- [ ] __Mock of Event#publish to be referentially transparent__
  https://github.com/rchain/rchain/blob/b5ba61/comm/src/test/scala/coop/rchain/p2p/EffectsTestInstances.scala#L129
  Move mock from _comm_ project (?).

- [ ] __Define _syntax_ trait for all extensions in _models_ project for easy import__
  Implicit class should extend `AnyVal` to prevent boxing.
  https://github.com/rchain/rchain/blob/003e2a8e/models/src/main/scala/coop/rchain/models/BlockHash.scala#L10-L12
  https://github.com/rchain/rchain/blob/af91c36d1/models/src/main/scala/coop/rchain/models/either/syntax/Either.scala#L9
- [x] __Remove or comment out unused code__
https://github.com/rchain/rchain/blob/f43905bd/rspace/src/main/scala/coop/rchain/rspace/history/HistoryRepository.scala#L20-L27 e.g.
- [ ] __Move ProtoUtil methods not related to casper into to models project. Check for other model logic in other projects__
https://github.com/rchain/rchain/blob/9ef40ea099e7d3328d485144a2d7c92b2000709b/casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala#L28
- [ ] __Cleanup all metrics: unify names, remove unused, add necessary__
- [ ] **Review the TODO comment in [Main.scala](https://github.com/rchain/rchain/blob/5d7372884/node/src/main/scala/coop/rchain/node/Main.scala#L89-L90) if it's still relevant because referenced JVM bug seems resolved.**
- [ ] **Rename the generated bnfc to `Parser` instead of `Syntax` based on the discussion** https://github.com/rchain/rchain/pull/3193#discussion_r519160300
https://github.com/rchain/rchain/blob/116c21bf647bc634251b13bd8021487663b560fb/project/BNFC.scala#L95
- [ ] **Rewrite using tailRecM** https://github.com/rchain/rchain/blob/dcd5ee6656adf572bd4963844a880d893593410a/rspace/src/main/scala/coop/rchain/rspace/SpaceMatcher.scala#L72
- [ ] **Rewrite code where `traverse` is used, make it concurrent using streams with parJoin** 
- [ ] **tailRecM?** https://github.com/rchain/rchain/blob/6dd06305020f02f9fed99c7eda6cbd980fdb0b7d/rholang/src/main/scala/coop/rchain/rholang/interpreter/Reduce.scala#L87

### PR DESCRIPTION
**Pull request by vinay-bhamidipati**
## Overview

Renames bnfc generated files from package syntax to ast to improve clarity. 
Described in #3001 and #3193 

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

File: project/BNFC.scala [hunk 1]
```diff
@@ -92,7 +92,7 @@ object BNFC {
         // format: off
         javaSource     := (javaSource in Compile).value,
         scalaSource    := (javaSource in Compile).value,
-        bnfcNamespace  := "coop.rchain.rholang.syntax",
+        bnfcNamespace  := "coop.rchain.rholang.parser",
```
Comment (tgrospic) [inline project/BNFC.scala]: @nzpr, @zsluedem I like _syntax_ more then _parser_ here, but we use syntax for extensions.
What do you think about `coop.rchain.rholang.ast`?

Comment (zsluedem) [reply project/BNFC.scala]: ast is Ok to me

Comment (zsluedem) [reply project/BNFC.scala]: parser is also ok to me.

Comment (tgrospic) [reply project/BNFC.scala]: @vb-cs please change _parser_ to `ast`. It's more precise meaning for types generated from BNFC grammar.

**Comment by tgrospic** [comment]
Now that you have resolved the name collision please also change the main issue that we had to declare syntax object in the rholang package. Discussion started here https://github.com/rchain/rchain/pull/3193#discussion_r519160300.

So syntax definition from interpreter package should ends up in the rholang package.
https://github.com/rchain/rchain/blob/6dd06305020f02f9fed99c7eda6cbd980fdb0b7d/rholang/src/main/scala/coop/rchain/rholang/interpreter/package.scala#L11-L14

https://github.com/rchain/rchain/blob/67c8144b038de9f5d3c7e8b8c9bdf348c089af1c/rholang/src/main/scala/coop/rchain/rholang/package.scala#L5

**Comment by vinay-bhamidipati** [comment]
> 
> 
> Now that you have resolved the name collision please also change the main issue that we had to declare syntax object in the rholang package. Discussion started here [#3193 (comment)](https://github.com/rchain/rchain/pull/3193#discussion_r519160300).
> 
> So syntax definition from interpreter package should ends up in the rholang package.
> 
> https://github.com/rchain/rchain/blob/6dd06305020f02f9fed99c7eda6cbd980fdb0b7d/rholang/src/main/scala/coop/rchain/rholang/interpreter/package.scala#L11-L14
> 
> https://github.com/rchain/rchain/blob/67c8144b038de9f5d3c7e8b8c9bdf348c089af1c/rholang/src/main/scala/coop/rchain/rholang/package.scala#L5

I see. I will work on this tomorrow.

**Comment by zsluedem** [comment]
This is a pr wanted to be merged into feature/block-merge branch. The tests in  feature/block-merge are broken because of we merged with tests broken before https://github.com/rchain/rchain/pull/3201.
@vb-cs You should pr to `dev` branch

**Comment by tgrospic** [comment]
@zsluedem I was mistaken for `RhoRuntimeSyntax`. It's not part of dev branch so @vb-cs rebase PR to block merge branch. We already discussed this to rebase it back to dev.

**Review by zsluedem — APPROVED** [review]
LGTM. https://github.com/rchain/rchain/runs/1633835972?check_suite_focus=true tests were made to fail here.

**Pull request closed without merging**